session.js

```diff
@@ -0,0 +1,111 @@
+/*!
+ * diva-session
+ * Copyright(c) 2017 Alliander, Koen van Ingen, Timen Olthof
+ * BSD 3-Clause License
+ */
+
+const divaStateModule = require('./diva-state');
+
+let divaState;
+
+/**
+* Module dependencies.
+* @private
+*/
+
+const BPromise = require('bluebird');
+
+function mergeAttribute(attributes, attributeName, attributeValue) {
+  // If an attribute of the same type is already stored, add the
+  // new value. Otherwise, add a new array with the first value.
+  const valuesForAttributeName = attributes[attributeName] ?
+    attributes[attributeName].concat(attributeValue) :
+    [attributeValue];
+  return {
+    ...attributes,
+    [attributeName]: valuesForAttributeName,
+  };
+}
+
+function getAttributes(divaSessionId) {
+  return divaState.getDivaEntry(divaSessionId)
+    .then((divaStateEntry) => {
+      let attributes = {};
+      Object.values(divaStateEntry).forEach((proof) => {
+        if (proof.status === 'VALID') {
+          const attributeMap = proof.attributes;
+          Object.keys(attributeMap).forEach((name) => {
+            attributes = mergeAttribute(attributes, name, attributeMap[name]);
+          });
+        }
+      });
+
+      return attributes;
+    });
+}
+
+function getMissingAttributes(divaSessionId, requiredAttributes) {
+  return getAttributes(divaSessionId)
+    .then((attributes) => {
+      const existingAttributes = Object.keys(attributes);
+      return requiredAttributes.filter(el => !existingAttributes.includes(el));
+    });
+}
+
+function addIrmaProofToSession(proofResult, irmaSessionId) {
+  const divaSessionId = proofResult.jti;
+
+  return divaState.getDivaEntry(divaSessionId)
```

Comment:
I think this should be separated from divaState even more.

Reply:
Agreed, however, let's do this in a later PR?

Reply:
ok